src/viper/README.md

```diff
@@ -161,23 +163,121 @@ Below, we summarize the language features that _Motoko-san_ currently supports.
+
+    * Relations on `Bool`, `Int`, `Nat`
+        * `==`, `!=`, `<`, `>`, `>=`, `<=`
+
+    * Array types `[T]` and `[var T]`, where `T` ∈ {`Bool`, `Int`, `Nat`}
+        * Construction: `[]`, `[x]`, `[x, y, z]`, `[var x, y, z]`, etc
+        * Indexing: `a[i]`, can be used on the LHS of an assignment
+        * Operations: `a.size()`
+
+    * Tuple types `(T₁, T₂)`, `(T₁, T₂, T₃)`, etc, where `T` ∈ {`Bool`, `Int`, `Nat`}
+        * Construction: `(a, b)`, `(a, b, c)`, etc
+        * Projections: `x.1`, `x.2`, etc
+        * Deconstruction by pattern matching:
+
+            * `let (a, b) = x`
+            * `switch x { case (a, b) ... }`
+
+          Nested matching is _not_ supported, i.e. subpatterns for tuple
+          elements `a`, `b`, `c` in `(a, b, c)` must all be variable patterns.
+
+    * Option types `Option<T>` (no restrictions on `T`)
+        * Construction: `null`, `?a`
+        * Deconstruction by pattern matching:
+
+          ```motoko
+          switch x {
+              case null ...;
+              case (?a) ...;
+          }
+          ```
+
+          Nested matching is _not_ supported, i.e. the subpattern `a` in `?a`
+          must be a variable pattern.
+
+    * Variant types `{ #A; #B : T₁; #C : (T₂, T₃) }` (with limitations)
+        * Construction: `#A`, `#B(x)`, `#C(x, y)`
+        * Deconstruction by pattern matching:
 
-  * `Int`: `+`, `/`, `*`, `-`, `%`
+          ```motoko
+          switch x {
+              case #A ...;
+              case #B(x) ...;
+              case #C(x, y) ...;
+          }
+          ```
 
-  * Relations: `==`, `!=`, `<`, `>`, `>=`, `<=`
+          Nested matching is _not_ supported, i.e. the subpatterns `x`, `y` in
+          `#B(x)` and `#C(x, y)` must all be variable patterns.
 
-  Supporting `Text`, `Nat`, `Int32`, tuples, record, and variants is _simple_.
+        * Generic and recursive variants are supported, e.g.
 
-    Supporting `Option<T>` types is _trivial_.
+          ```motoko
+          type List<T> = { #Nil; #Cons : (T, List<T>) }
+          ```
+
+        * Limitation 1: Nominative equality. In the Viper translation, equality
+          of variant types is not _structural_ (as in normal Motoko programs)
+          but _nominative_.
+          * One consequence of this is that a variant type may not be used
+            directly but must be bound to a name:
+
+            ```motoko
+            type N = { #A; #B }
+            ```
+
+            Instead of using `{ #A; #B }` in type signatures, one must always
+            refer to this variant by its name `N`.
+          * Furthermore, given two identical variant declarations, they are
+            treated as distinct in the translation:
+
+             ```motoko
+             type N = { #A; #B }
+             type M = { #A; #B }   // M ≠ N
+             ```
+
+          * Motoko's type checker has not been modified to account for
+            nominative equality of variants, so type inference is not reliable:
```

Comment:
```suggestion
            nominal equality of variants, so type inference is not reliable:
```
I guess we will just have to live with this fundamental difference between Viper and Motoko. @aterga is there any way to teach Viper about structural type equivalence (and subtyping)?

Reply:
The Viper language is designed to be simple, and to allow modeling programs in many different (imperative) languages. So while we cannot translate many statement from Motoko into Viper 1:1, of course there are multiple ways we could model those cases in Viper and still support them. 

That is where I was going with this comment, btw: https://github.com/serokell/motoko/pull/19#discussion_r1616893231

Reply:
For example [Prusti](https://www.pm.inf.ethz.ch/research/prusti.html) encode all Rust types as predicates. I think in that approach subtyping is not a big deal, but there will be other drawbacks e.g. folding/unfolding Viper isorecursive predicates.

Reply:
Correct. I would not recommend using separation logic predicates in Motoko-san, as they would require generating ghost statements to fold / unfold the definitions, and they are limited to tree structures (so, e.g., an array cannot be expressed this way).